### ISSUE TITLE
bug: Upgrade jest to v28

### ISSUE THREAD

**Issue by Tallyb** [ionic-team/stencil#3348]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

2.15.1

### Current Behavior

Errors when upgrading jest to v28

### Expected Behavior

Stencil to work with jest 28 after installing peer deps

### Steps to Reproduce

Run npm t on the attached repo (branch jest-28)
you can see the errors here: https://github.com/Tallyb/stencil-one/actions/runs/2247154163

### Code Reproduction URL

https://github.com/Tallyb/stencil-one/tree/jest-28

**Comment by rwaskiewicz** [comment]
Hey @Tallyb 👋 

You're correct that Jest 28 isn't compatible with Stencil at the moment.  At this time, Stencil doesn't support Jest 28, although it's something we'll be working on in the future.  We will likely be working through migrating Stencil internally to jest-circus, internally moving to Jest 28, then adding support for end users.  For full transparency, this is not currently at the top of our priority list.  The good news is that we've learned quite a bit from adding Jest 27 support, and hope to leverage that knowledge to expedite implementing Jest 28 support when the time comes.

In the next release of Stencil, we will be shipping an update to our CLI that makes it clearer to folks what versions of Jest we do and don't support. Today, installing a new major version of Jest and running a test script is a bit of a gamble - will it work? will it not? We hope that we can be a little clearer about what we do/don't support in the CLI so that folks aren't greeted with a stack trace when trying to run their tests.

**Comment by Tallyb** [comment]
Well, specifying supporting version is far from being a solution to the issue. With the ecosystem moving forward, it is no use being left behind. 
Is Stenciljs a dead project by now?

**Comment by ghaiat** [comment]
Hello, any update on this? Angular 14 has jest 28, we cannot uprade until this is fixed. it s been opened for quite a while....

**Comment by rwaskiewicz** [comment]
Hey folks,

At this time, we don’t have anything new to share on Jest 28 support, other than it is still on our radar.

Stencil is still very much an active project, that works with the open source community, the Ionic Framework, Ionic’s Enterprise customers and our internal engineering team to accept & prioritize work on the project.  We ask that folks be patient while we balance the different asks we receive from these various groups.

Thanks for your understanding,

The Stencil Team

**Comment by Tallyb** [comment]
For those interested, here is a hack for working with test specs (not with e2e) that you can use meanwhile: 
1) If you are using stencil spec -> extract the stencil configuration to a jest.config.ts file. Your jest config should include

**Comment by daragh-king-genesys** [comment]
@rwaskiewicz any update on this? Is it possible yet to estimate when this will get prioritised?

**Comment by maidi29** [comment]
@rwaskiewicz any update on this?
Would be cool to use @GlennLefeveres fix

**Comment by dgonzalezr** [comment]
@rwaskiewicz this is definitely a blocker for those using Nx and wanting to migrate to their latest version 😞 . We are stuck with Nx 14 while knowing that Nx 15 shifts a lot of new stuff and improvements that we cannot get the advantage of.

**Comment by ghaiat** [comment]
hello. any updates here? we re early 2023...

**Comment by sean-perkins** [comment]
@ghaiat the team is regrouping after delivering the new major release of Stencil last week (v3.0.0). 

Ryan's comment here is still the best reflection of where this issue stands: https://github.com/ionic-team/stencil/issues/3348#issuecomment-1340958144

The team has designed the work behind adding Jest 28+ support and is prioritizing the roadmap for the upcoming year. The team will share an update when that work is available for testing or released to the public. 

Thanks!

**Comment by Tallyb** [comment]
PAtch for stencil 3.0: 
```
diff --git a/node_modules/@stencil/sass/.DS_Store b/node_modules/@stencil/sass/.DS_Store
new file mode 100644
index 0000000..e69de29
diff --git a/node_modules/@stencil/sass/dist/index.js b/node_modules/@stencil/sass/dist/index.js
index 42bda03..c989ed2 100644
--- a/node_modules/@stencil/sass/dist/index.js
+++ b/node_modules/@stencil/sass/dist/index.js
@@ -248,7 +248,8 @@ function getRenderOptions(opts, sourceText, fileName, context) {
                 }
             }
             const importTerminator = renderOpts.indentedSyntax ? '\n' : ';';
-            return `@import "${injectGlobalPath}"${importTerminator}`;
+            return `@use "${injectGlobalPath}" as *${importTerminator}`;
+            //return `@import "${injectGlobalPath}"${importTerminator}`;
         })
             .join('');
         renderOpts.data = injectText + renderOpts.data;
diff --git a/node_modules/@stencil/sass/dist/util.js b/node_modules/@stencil/sass/dist/util.js
index 24590e2..d3edece 100644
--- a/node_modules/@stencil/sass/dist/util.js
+++ b/node_modules/@stencil/sass/dist/util.js
@@ -59,7 +59,8 @@ export function getRenderOptions(opts, sourceText, fileName, context) {
                 }
             }
             const importTerminator = renderOpts.indentedSyntax ? '\n' : ';';
-            return `@import "${injectGlobalPath}"${importTerminator}`;
+            return `@use "${injectGlobalPath}" as *${importTerminator}`;
+            //return `@import "${injectGlobalPath}"${importTerminator}`;
         })
             .join('');
         renderOpts.data = injectText + renderOpts.data;
```

**Comment by sebastiandg7** [comment]
@Tallyb With the v3 release, do you think we are close to having Jest 28+ available?

**Comment by sarahquigley** [comment]
Any news on this? Stencil is lagging several versions behind at this point, and it would be great to be able to upgrade. Particularly causes issues in an NX setting.

**Comment by theo-staizen** [comment]
@Tallyb do you have a fork with your fixes?
or are you using [patch-package](https://www.npmjs.com/package/patch-package) to patch stencil code?

**Comment by Tallyb** [comment]
Patch-package

**Comment by sarahquigley** [comment]
This issue has been open for over a year now and the supported jest version is very out of date. Any timeline on supporting latest jest?

**Comment by katja-brunner** [comment]
Updating to Stencil 3.3.1 allows Typescript update to 5.1.3. But Typescript 5+ is not accepted by ts-jest 27+, so it is required to update ts-jest to 29+, what requires all jest dependencies to update to 29+. But as Stencil is still not supporting jest 29+ we cannot use latest Typescript. So now we have another dependency we cannot update due to this bug. This is very annoying. When will there be a solution?

**Comment by rwaskiewicz** [comment]
Hey folks 👋 

Support for newer versions of Jest will be our next item to tackle. We made the decision to remove additional technical debt/unused Stencil functionality in favor of Jest support this past spring. Please keep in mind, this was not a decision that was taken lightly.

---

Our plan as of this writing is to begin supporting Jest 29 in the near future. Our desire is for this to be the first step in decoupling Jest from Stencil. Today, Stencil's [supported testing stack](https://stenciljs.com/docs/support-policy#testing-libraries) includes:
- Jest 24 thru 27
- Puppeteer 10 thru 20

However, we are interested in hearing what folks are using today outside of this stack/why. We're also interested in hearing what folks would like to use and why. To that end, I've created 🧪 [Stencil Testing Packages Infrastructure](https://github.com/ionic-team/stencil/discussions/categories/stencil-testing-packages-infrastructure) for folks to add their feedback.

**Comment by rwaskiewicz** [comment]
> Updating to Stencil 3.3.1 allows Typescript update to 5.1.3. But Typescript 5+ is not accepted by ts-jest 27+, so it is required to update ts-jest to 29+, what requires all jest dependencies to update to 29+. But as Stencil is still not supporting jest 29+ we cannot use latest Typescript. So now we have another dependency we cannot update due to this bug. This is very annoying. When will there be a solution?

@katja-brunner 

Please note that Stencil does not support TypeScript 5.1 at this time. We do not support/integrate directly with ts-jest either. We'd be super interested in hearing more about this use case  - would you mind opening a discussion in 🧪 [Stencil Testing Packages Infrastructure](https://github.com/ionic-team/stencil/discussions/categories/stencil-testing-packages-infrastructure) so we could learn more

**Comment by sarahquigley** [comment]
Is there any specific timeline for this work as yet?

**Comment by rwaskiewicz** [comment]
The fix for this issue has been released as a [part of today's Stencil v4.7.0 release](https://github.com/ionic-team/stencil/releases/tag/v4.7.0). Should the issue reappear, please feel free to open a new ticket. Thanks!

**Comment by mjsorribas** [comment]
Hi! I've had many issues with the latest version of Stencil 4.22.2 and the jest version. I recommend creating a compatibility matrix and publishing it as part of the documentation. Also with NX as the monorepo, the issues with Jest 29 still there. I. can't use the last NX because use Jest 29.x, but Stencil is not compatible. And we are at the end of Q3 2024. The fix was for 29.0.0 :
```sys.lazyRequire = new NodeLazyRequire(nodeResolve, {
    '@types/jest': { minVersion: '24.9.1', recommendedVersion: '29', maxVersion: '29.0.0' },
    jest: { minVersion: '24.9.0', recommendedVersion: '29', maxVersion: '29.0.0' },
    'jest-cli': { minVersion: '24.9.0', recommendedVersion: '29', maxVersion: '29.0.0' },
    puppeteer: { minVersion: '10.0.0', recommendedVersion: '20' },
    'puppeteer-core': { minVersion: '10.0.0', recommendedVersion: '20' },
    'workbox-build': { minVersion: '4.3.1', recommendedVersion: '4.3.1' },
  });```

The current Jest version is 29.7.0.
Without a matrix of compatibility, upgrading any package is a potential issue.

**Comment by rwaskiewicz** [comment]
@mjsorribas 

Hey there 👋 

I recommend opening a new issue with a minimal reproduction case for the team to look into. Stencil v4 is meant to be compatible with all minor versions of Jest v29, but it's possible there's something amiss here. When doing so, please add as much detail as possible - including the exact steps to reproduce the problem, what you've tried so far to investigate the issue, etc. This helps the team triage issues effectively.

For what it's worth, the team does have a compatibility matrix for Stencil and Jest, which can be found at https://stenciljs.com/docs/support-policy#jest
